cpp/include/cudf/detail/join.hpp

```diff
@@ -0,0 +1,186 @@
+
+ public:
+  /**
+   * @brief Constructor that internally builds the hash table based on the given `build` table.
+   *
+   * @throw cudf::logic_error if the number of columns in `build` table is 0.
+   * @throw cudf::logic_error if the number of rows in `build` table exceeds MAX_JOIN_SIZE.
+   *
+   * @param build The build table, from which the hash table is built.
+   * @param compare_nulls Controls whether null join-key values should match or not.
+   * @param stream CUDA stream used for device memory operations and kernel launches.
+   */
+  hash_join(cudf::table_view const& build,
+            cudf::null_equality compare_nulls,
+            rmm::cuda_stream_view stream = rmm::cuda_stream_default);
+
+  /**
+   * @copydoc cudf::hash_join::inner_join
+   */
+  std::pair<std::unique_ptr<rmm::device_uvector<size_type>>,
+            std::unique_ptr<rmm::device_uvector<size_type>>>
+  inner_join(cudf::table_view const& probe,
+             std::optional<std::size_t> output_size,
+             rmm::cuda_stream_view stream,
+             rmm::mr::device_memory_resource* mr) const;
+
+  /**
+   * @copydoc cudf::hash_join::left_join
+   */
+  std::pair<std::unique_ptr<rmm::device_uvector<size_type>>,
+            std::unique_ptr<rmm::device_uvector<size_type>>>
+  left_join(cudf::table_view const& probe,
+            std::optional<std::size_t> output_size,
+            rmm::cuda_stream_view stream,
+            rmm::mr::device_memory_resource* mr) const;
+
+  /**
+   * @copydoc cudf::hash_join::full_join
+   */
+  std::pair<std::unique_ptr<rmm::device_uvector<size_type>>,
+            std::unique_ptr<rmm::device_uvector<size_type>>>
+  full_join(cudf::table_view const& probe,
+            std::optional<std::size_t> output_size,
+            rmm::cuda_stream_view stream,
+            rmm::mr::device_memory_resource* mr) const;
+
+  /**
+   * @copydoc cudf::hash_join::inner_join_size
+   */
+  [[nodiscard]] std::size_t inner_join_size(cudf::table_view const& probe,
+                                            rmm::cuda_stream_view stream) const;
+
+  /**
+   * @copydoc cudf::hash_join::left_join_size
+   */
+  [[nodiscard]] std::size_t left_join_size(cudf::table_view const& probe,
+                                           rmm::cuda_stream_view stream) const;
+
+  /**
+   * @copydoc cudf::hash_join::full_join_size
+   */
+  std::size_t full_join_size(cudf::table_view const& probe,
+                             rmm::cuda_stream_view stream,
+                             rmm::mr::device_memory_resource* mr) const;
+
+ private:
+  /**
+   * @brief Probes the `_hash_table` built from `_build` for tuples in `probe_table`,
+   * and returns the output indices of `build_table` and `probe_table` as a combined table,
+   * i.e. if full join is specified as the join type then left join is called. Behavior
+   * is undefined if the provided `output_size` is smaller than the actual output size.
+   *
+   * @throw cudf::logic_error if hash table is null.
```

Comment:
There appears to be a special case for this
https://github.com/rapidsai/cudf/pull/10695/files#diff-7e4a3b3c2c7b9cf3ae4c5ef82cd665e8b3579e0599dd3536c754424b000a6e0eR385-R388

Also, does `null == empty` here?  Generally they mean different things. There is an `_is_empty` variable that appears to match with `is null` statements in the documentation.